tests/integrationv2/test_serialization.py

```diff
@@ -0,0 +1,160 @@
+import pytest
+import copy
+import os
+from enum import Enum, auto
+
+from configuration import available_ports
+from common import ProviderOptions, Protocols, random_str
+from fixtures import managed_process  # lgtm [py/unused-import]
+from providers import Provider, S2N
+from utils import invalid_test_parameters, get_parameter_name, to_bytes
+
+SERVER_STATE_FILE = 'server_state'
+CLIENT_STATE_FILE = 'client_state'
+
+SERVER_DATA = f"Some random data from the server:" + random_str(10)
+CLIENT_DATA = f"Some random data from the client:" + random_str(10)
+
+
+class MainlineRole(Enum):
+    Serialize = auto()
+    Deserialize = auto()
+
+
+"""
+This test file checks that a serialized connection can be deserialized by an older version of
+s2n-tls and vice versa. This ensures that any future changes we make to the handshake are backwards-compatible
+with an older version of s2n-tls.
+
+This feature requires an uninterrupted TCP connection with the peer in-between serialization and
+deserialization. Our integration test setup can't easily provide that while also using two different
+s2n-tls versions. To get around that we do a hack and serialize/deserialize both peers in the TLS connection.
+This prevents one peer from receiving a TCP FIN message and shutting the connection down early.
+"""
+
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13, Protocols.TLS12], ids=get_parameter_name)
+@pytest.mark.parametrize("mainline_role", [MainlineRole.Serialize, MainlineRole.Deserialize], ids=get_parameter_name)
+def test_server_serialization_backwards_compat(managed_process, tmp_path, protocol, mainline_role):
```

Comment:
Oh yeah the enum is much clearer! Nice.

I wonder if you need separate code for the client vs server though. They're basically exactly the same function, except that one turns use_mainline_version on for the client and one does it for the server. Wouldn't it be easier to just use a parameter to choose which?